src/transformStateWithClones.js

```diff
@@ -7,7 +7,52 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let previousState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        previousState = createStateWithAddedProperties(
+          previousState,
+          action.extraData,
+        );
+        break;
+      case 'removeProperties':
+        previousState = createStateWithRemovedProperties(
+          previousState,
+          action.keysToRemove,
+        );
+        break;
+      case 'clear':
+        previousState = createClearState(previousState);
```

Comment:
The function `createClearState` is misleading as it suggests that the state is being cleared when in fact a new empty object is being returned. A clearer name would be `createEmptyStateCopy`.